File: Marcianitos/src/deepspace/SpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author laura
 */
public class SpaceStation {
    static private int MAXFUEL = 100;     //atributo de clase creo q es asi
    static private float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;   
    private Damage pendingDamage;
    
    private void assignFuelValue(float f){
        if (f < MAXFUEL)
            fuelUnits = f;
        else
            fuelUnits = MAXFUEL;
    }
    
    private void cleanPendingDamage(){
        if (getPendingDamage().hasNoEffect())
            pendingDamage = null;        
    }
    
    SpaceStation(String n, SuppliesPackage supplies){
        this.name = n;
        
        // creo q estos se inicializarian a 0 pq si no estamos sumando cosas a algo q no tiene valor
        this.ammoPower = 0;
        this.shieldPower = 0;
        this.fuelUnits = 0;
        
        this.receiveSupplies(supplies);
        
        // estos de abajo creo q se inicializarian así
        this.nMedals = 0; 
        this.weapons = null;
        this.shieldBoosters = null;
        this.hangar = null;
        this.pendingDamage = null;       
    }
    
    public void cleanUpMountedItems() {
        // Compruebo las armas sin usos y elimino las que no le queden usos
        int num_weapons = this.weapons.size();
        for (int i = 0; i < num_weapons; ++i){
            if (weapons.get(i).getUses() == 0){
                weapons.remove(i);
            }
        }

        //Compruebo los escudos sin usos y elimino los que no tengan usos
        int num_shields = this.shieldBoosters.size();
        for (int i = 0; i < num_shields; ++i){
            if (this.shieldBoosters.get(i).getUses() == 0){
                this.shieldBoosters.remove(i);
            }

        }
    }
    
    public void discardHangar() {
        this.hangar=null;        
    }
    
    public void discardShieldBooster(int i){
        throw new UnsupportedOperationException();
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (getHangar() != null)
            getHangar().removeShieldBooster(i);
    }
    
    public void discardWeapon(int i){
        //throw new UnsupportedOperationException();
        
        int size = weapons.size();
        
        if(i>=0 && i<size){
            
            Weapon w = weapons.remove(i);
            
            if(pendingDamage != null){
                
                pendingDamage.discardWeapon(w);
                
                cleanPendingDamage();
            }
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (getHangar() != null)
            getHangar().removeWeapon(i);
    }
    
    public float fire(){
<<<<<<< HEAD
        
        // throw new UnsupportedOperationException();
        
        int size = weapons.size();
        
        float factor=1f;
        
        Weapon w;
        
        for(int i=0;i<size;i++){
            
            w = weapons.get(i);
            
            factor *= w.useIt();
                      
        }
        
        return (ammoPower*factor);           
          
=======
        int size = this.weapons.size();
		
		float factor = 1f;

		for (int i = 0; i < size; ++i){
			Weapon w = weapons.get(i);
			factor *= w.useIt();
		}
		
		return (ammoPower*factor);
>>>>>>> 67ab684375894aa35324610b685f40107b439ebe
    }
    
    
    
    public float getAmmoPower(){
        return this.ammoPower;
    }
    
    public float getFuelUnits(){
        return this.fuelUnits;
    }
    
    public Hangar getHangar(){
        return this.hangar;
    }
    
    public String getName(){
        return this.name;
    }
    
    public int getNMedals(){
        return this.nMedals;
    }
    
    public Damage getPendingDamage(){
        return this.pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return this.shieldBoosters;
    }
    
    public float getShieldPower(){
        return this.shieldPower;
    }
    
    public float getSpeed(){
        return (this.fuelUnits/SpaceStation.MAXFUEL);
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons(){
        return this.weapons;
    }
    
    public void mountShieldBooster(int i){
        ShieldBooster s;
        if (getHangar() != null) {
            s = getHangar().removeShieldBooster(i);
            if (s != null)
                getShieldBoosters().add(s);
        }        
    }
    
    public void mountWeapon(int i){
        Weapon w;
        if (getHangar() != null) {
            w = getHangar().removeWeapon(i);
            if (w != null)
                getWeapons().add(w);
        }
    }
    
    public void move(){
        this.fuelUnits -= this.fuelUnits*this.getSpeed();
    }
    
    public float protection(){
        int size = this.shieldBoosters.size();
        int factor = 1;
        for (int i = 0; i < size; ++i) {
            ShieldBooster s = this.shieldBoosters.get(i);
            factor *= s.useIt();
        }
        return (shieldPower * factor);
    }
    
    public void receiveHangar(Hangar h){
        if (getHangar() == null)    
            this.hangar = h;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        boolean resultado = false;
        
        if (getHangar() != null)  
            resultado = getHangar().addShieldBooster(s);
        
        return resultado; 
    }
    
    public ShotResult receiveShot(float shot){
        throw new UnsupportedOperationException();
    }
    
    public void receiveSupplies(SuppliesPackage s){
        this.ammoPower += s.getAmmoPower();
        this.shieldPower += s.getshieldPower();
        this.assignFuelValue(this.fuelUnits + s.getfuelUnits());
    }
    
    public boolean receiveWeapon(Weapon w){
        boolean resultado = false;
        
        if (getHangar() != null)  
            resultado = getHangar().addWeapon(w);
        
        return resultado;
    }
    
	public void setLoot(Loot loot) {
        CardDealer dealer = CardDealer.getInstance();
		
		// Aniadimos Hangars
		int h = loot.getNHangars();
		if (h > 0){
			hangar = dealer.nextHangar();
			
			this.receiveHangar(hangar);
		}
		
		// Aniadimos supplies
		int elements = loot.getNSupplies();
		for (int i = 1; i < elements; ++i){
			SuppliesPackage sup = dealer.nextSuppliesPackage();
			receiveSupplies(sup);
		}
		
		// Aniadimos Weapons
		elements = loot.getNWeapons();
		for (int i = 1; i < elements; ++i){
			Weapon weap = dealer.nextWeapon();
			receiveWeapon(weap);
		}
		
		// Aniadimos Shields
		elements = loot.getNShields();
		for (int i = 1; i < elements; ++i){
			ShieldBooster sh = dealer.nextShieldBooster();
			
			receiveShieldBooster(sh);
		}
		
		int medals = loot.getNMedals();
		
		this.nMedals += medals;
    }
    
    public void setPendingDamage(Damage d){
        d.adjust(weapons, shieldBoosters);
        // se almacena el resultado en el atributo correspondiente ???
        // creo q es esto:
        this.pendingDamage = d;
    }
    
    public boolean validState(){
        return (this.pendingDamage == null || this.pendingDamage.hasNoEffect());
    }
}

File: RUBY/src/SpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStationToUI'
require_relative 'Weapon'
require_relative 'WeaponType'
require_relative 'ShieldBooster'
require_relative 'Hangar'
require_relative 'Damage'
require_relative 'SuppliesPackage'
require_relative 'ShotResult'

module Deepspace
  
	class SpaceStation

		#atributos de clase
		@@MAXFUEL=100
		@@SHIELDLOSSPERUNITSHOT=0.1
	  
		attr_reader :name, :nMedals, :ammoPower, :fuelUnits, \
		  :shieldPower, :shieldBoosters, :hangar, :weapons, \
		  :pendingDamage, :MAXFUEL, :SHIELDLOSSPERUNITSHOT
	  
		def initialize (n, supplies)		
			@name=n
			@ammoPower = 0
			@shieldPower = 0
			@fuelUnits = 0
			receiveSupplies(supplies)
			
			@nMedals=0
			@weapons=Array.new()
			@shieldBoosters=Array.new()
			@hangar=nil
			@pendingDamage=nil
		end
		
		def cleanUpMountedItems()
			#Compruebo las armas sin usos y elimino las que no le queden usos
			@weapons.each do |w|
				if (w.uses == 0)
					@weapons.delete(w) #creo q es asi
				end
			end

			#Compruebo los escudos sin usos y elimino los que no tengan usos
			@shieldBoosters.each do |s|
				if (s.uses == 0)
					@shieldBoosters.delete(s) #creo q es asi
				end
			end
		end
		
		def discardHangar()
			@hangar=nil
		end
		
		def discardShieldBooster(i)
			size = @shieldBoosters.length
			if (i>=0 && i<size)
				s=@shieldBoosters.delete_at(i)
				
				if (@pendingDamage != nil)
					@pendingDamage.discardShieldBooster(w)
					cleanPendingDamage
				end
			end	
		end
		
		def discardShieldBoosterInHangar(i)
			if (hangar != nil)
				hangar.removeShieldBooster(i) 
			end
		end
		
		def discardWeapon(i)
		
			size=@weapons.length
			if (i>=0 && i<size)
				w=@weapons.delete_at(i)
				
				if (@pendingDamage != nil)
					@pendingDamage.discardWeapon(w)
					cleanPendingDamage
				end
			end		
		end
		
		def discardWeaponInHangar(i)
			if (hangar != nil)
				hangar.removeWeapon(i)
			end
		end
		
		def fire()
			factor = 1
			
			@weapons.each do |w|
				factor *= w.useIt()
			end
			
			return (ammoPower*factor)
		
		end
		
		def getNMedals()
			return @nMedals
		end
		
		def getSpeed() 
			return (@fuelUnits.to_f/@@MAXFUEL)
		end
		
		def getUIversion()
			SpaceStationToUI.new(self)
		end
		
		def mountShieldBooster(i)
			if (hangar != nil) 
				s = hangar.removeShieldBooster(i)
				if (s != nil)
					shieldBoosters.push(s)
				end
			end
		end
		
		def mountWeapon(i)
			if (hangar != nil) 
				w = hangar.removeWeapon(i)
				if (w != nil)
					weapons.push(w)
				end
			end
		end
		
		def move()
			@fuelUnits -= @fuelUnits*getSpeed()
		end
		
		def protection() 
			factor = 1
			shieldBoosters.each do |s|
				factor *= s.useIt
			end	
			return (@shieldPower * factor)		
		end
		
		def receiveHangar(h)
			if (hangar == nil)
				@hangar = h
			end
		end
		
		def receiveShieldBooster(s) 
			retorno = false
			
			if (hangar != nil)
				retorno = hangar.addShieldBooster(s)	#creo q asi devuelve true o false ya
			end
			
			return (retorno)
		end
		
		def receiveShot(shot)
			myProtection=protection
			if(myProtection>=shot)
				@shieldPower=@shieldPower-(@@SHIELDLOSSPERUNITSHOT*shot)
				@shieldPower=[0.0,shieldPower].max
				ShotResult::RESIST
			else
				@shieldPower=0.0
				ShotResult::DONOTRESIST
			end
		
		end
		
		def receiveSupplies(s) 
			@ammoPower += s.ammoPower
			@shieldPower += s.shieldPower
			
			assignFuelValue(fuelUnits + s.fuelUnits)
			
		end
		
		def receiveWeapon(w)
			retorno = false
			if (hangar != nil)
				retorno = hangar.addWeapon(w)	#creo q asi devuelve true o false ya
			end
		end
		
		def setLoot(loot)
			dealer = CardDealer.new()
			
			#Aniadimos Hangars
			h = @loot.getNHangars()
			if (h > 0)
				hangar = dealer.nextHangar()
				
				receiveHangar(hangar)
			end
			
			#Aniadimos supplies
			elements = @loot.getNSupplies()
			for i in (1..elements) do
				sup = dealer.nextSuppliesPackage()
				receiveSupplies(sup)
			end
			
			# Aniadimos Weapons
			elements = @loot.getNWeapons()
			for i in (1..elements) do
				weap = dealer.nextWeapon()
				receiveWeapon(weap)
			end
			
			#Aniadimos Shields
			elements = @loot.getNShields()
			for i in (1..elements) do
				sh = dealer.nextShieldBooster()
				receiveShieldBooster(sh)
			end
			
			medals = @loot.getNMedals

			@nMedals += medals

		end
		
		def setPendingDamage(d)
			d.adjust(weapons, shieldBoosters)
			# se almacena el resultado en el atributo correspondiente ???
			# creo q es esto:
			@pendingDamage = d
		end
		
		def validState()
			return (@pendingDamage == nil || @pendingDamage.hasNoEffect())
		end
		
		def to_s
			out="Space Station - Name: #{@name}\n"
			out+="\tnMedals: #{@nMedals}, Fuel units: #{@fuelUnits.round(2)}, Power: #{@ammoPower}, Shields: #{@shieldPower}\n"
			out+="\tWeapons: [#{@weapons.join(' // ')}]\n"
			out+="\tShieldBooster: [#{@shieldBoosters.join(' // ')}]\n"
			out+="\tHangars: #{@hangar}\n"
			out+="\tPendingDamage: #{@pendingDamage}\n" 
			out+="------- end of Space Station >> #{@name} << -------"
			return out
		end
		
		private 	#todos los metodos q haya debajo de esto son privados (solo los de instancia)
		def assignFuelValue(f)
			if (f < @@MAXFUEL)
				@fuelUnits = f
			else
				@fuelUnits = @@MAXFUEL
			end
		end
		
		def cleanPendingDamage()
			if (@pendingDamage.hasNoEffect())
				pendingDamage = nil 
			end 
		end

	end # class

	supplies = SuppliesPackage.new(1,10,3)
	estacion = SpaceStation.new(3, supplies)
	hangar = Hangar.new(3)
	shield = ShieldBooster.new("Escudo", 5, 2)

	weapon = Weapon.new("Arma de destruccion", WeaponType::LASER, 2)
	
	estacion.receiveHangar(hangar)

	estacion.receiveShieldBooster(shield)
	estacion.receiveWeapon(weapon)

	estacion.mountWeapon(0)

	puts estacion.protection

	puts estacion.to_s
	
end # module
